src/compiler/mir/typetable.rs

```diff
@@ -0,0 +1,584 @@
+
+            // The presence of this wildcard points indicates that the Type enum has variants that no longer make sense
+            // and should be refactored out.
+            _ => panic!("Invalid Type: {:?}", ty),
+        };
+
+        for idx in 0..self.table.len() {
+            if self.table[idx] == mir_ty {
+                return Some(TypeId(idx as u32));
+            }
+        }
+
+        None
+    }
+
+    /// Will return true if every Structure is [`MirStructDef::Defined`] and every referenced
+    /// [`TypeId`] is within the bounds of the table.
+    pub fn is_complete(&self) -> bool {
+        // Iterate through every entry in the table
+        for entry in &self.table {
+            if !self.is_defined(entry) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Returns true if the given type from this Type Table is defined. Defined means that
+    /// this [`TypeId`] exists in the table, any type it references exists in the table, and
+    /// (if it is a Structure) the structure is [`MirStructDef::Defined`] and not [`MirStructDef::Declared`].
+    fn is_defined(&self, ty: &MirTypeDef) -> bool {
+        let max_id = TypeId(self.table.len() as u32);
+        match ty {
+            MirTypeDef::Array { ty, .. } if *ty >= max_id => return false,
+            MirTypeDef::RawPointer { target, .. } if *target >= max_id => return false,
+            MirTypeDef::Structure { def, .. } if *def == MirStructDef::Declared => return false,
+            MirTypeDef::Structure {
+                def: MirStructDef::Defined(fields),
+                ..
+            } if fields.iter().any(|f| f.ty >= max_id) => return false,
+            _ => (),
+        }
+        true
+    }
+
+    /// Converts a [`Parameter`] to a [`Field`].
+    fn to_field(&mut self, p: &Parameter<SemanticContext>) -> Field {
+        let id = self.add(&p.ty);
+        Field {
+            name: p.name,
+            ty: id,
+        }
+    }
+}
+
+/// These are the most basic types available and correspond to the arithmetic and operand
+/// types avaible on the CPU. All other types in Bramble are constructed from these types.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum MirBaseType {
+    Unit,
+    Null,
+    U8,
+    U16,
+    U32,
+    U64,
+    I8,
+    I16,
+    I32,
+    I64,
+    F64,
+    Bool,
+    StringLiteral,
+}
+
+impl MirBaseType {
+    fn parse(ty: &Type) -> Option<Self> {
+        let base = match ty {
+            Type::Unit => Self::Unit,
+            Type::Null => Self::Null,
+            Type::U8 => Self::U8,
+            Type::U16 => Self::U16,
+            Type::U32 => Self::U32,
+            Type::U64 => Self::U64,
+            Type::I8 => Self::I8,
+            Type::I16 => Self::I16,
+            Type::I32 => Self::I32,
+            Type::I64 => Self::I64,
+            Type::F64 => Self::F64,
+            Type::Bool => Self::Bool,
+            Type::StringLiteral => Self::StringLiteral,
```

Comment:
Is it necessary to have the Base Types in the Type Table?  Could an alternative design be created where only constructed types are in the table?

For example:
```rust
enum TypeId {
    Base(BaseType),
    Constructed(u32),
```

This would make it impossible for the base types to not be included in the type table, while also allowing `TypeId` to implement the `Copy` trait.

The cost is that TypeId values are bigger in size. Comparisons may also be more costly.

For memory size, this suggestion would be 8 bytes vs 4 bytes for a u32.

For cost of comparisons, here is a Godbolt link showing how the comparison for this enum optimizes: https://godbolt.org/z/7dqWszfbq. Note that it is significantly more complex than comparing two integers. The two paths execute 8 instructions and one path executes 3 instructions.

Reply:
One thing to note, is that this design only affects the internal implementation of `TypeTable`. Users of type table cannot see the implementation of `TypeId` and therefore neither design choice will have any impact on the readability or maintainability of the MIR code which makes use of `TypeTable`. This means that the value of readability are much smaller.

Given this, and the fact that the readability improvement is relatively small, I have chosen to continue with the current design. It has many benefits for memory and execution performance and is only slightly more complex than the proposed design.